### PR TITLE
elf: support kernel module kfunc calls

### PR DESCRIPTION
**Pull request by shun159**
This commit adds support for kernel module function call support.

The fdArray parameter is used during BPF program load to pass module BTFs referenced by the program. ins.Offset is set to index into this array, and ins.Constant is assigned a BTFID in the BTF (of module or vmlinux). 

cf: https://lore.kernel.org/bpf/20211002011757.311265-7-memxor@gmail.com/

### CODE REVIEW

File: linker.go [hunk 1]
```diff
@@ -210,12 +210,6 @@ func fixupKfuncs(insns asm.Instructions) error {
 	return nil
 
 fixups:
-	// only load the kernel spec if we found at least one kfunc call
```
Comment (rgo3) [inline linker.go:213]: Keep this here and pass the kernelSpec to the `findKfuncInKernel()` function.

**Review by rgo3 — CHANGES_REQUESTED** [review]
Thanks for getting started with this @shun159!
I gave this a first pass and left some comments. From a high-level perspective I think this needs some changes:

* With the current kfunc implementation we detect kfunc calls in the elf_reader, set some metadata on the instruction but resolve btf ids in the linker. In my opinion there is no need to change the elf_reader for this PR. The idea for the elf_reader is to deterministically produce a CollectionSpec on different systems [[1]](https://github.com/cilium/ebpf/blob/master/ARCHITECTURE.md#elf), starting to lookup types in the kernel or module specs would interfere with that.
* We don't need to parse `kallsyms`. Without it the patch becomes much shorter and easier to review going forward. It should be enough to iterate over the btf handles similarly as we do in findTargetInKernel:
https://github.com/cilium/ebpf/blob/fdcdde5d27ce7553cfa0986ea4f4ec1ea3be99ea/prog.go#L890-L900
The way I would do this would look something like this:
```go
func findKfuncInKernel(kernelSpec *btf.Spec, kfunc *btf.Func) (*btf.Handle, btf.TypeID, error) {
	var fn *btf.Func
	err := kernelSpec.TypeByName(kfunc.Name, &fn)
	if errors.Is(err, btf.ErrNotFound) {
		// couldn't find kfunc in vmlinux, searching modules.
		module, id, err := findTargetInModule(kfunc.Name, fn)
		if err != nil {
			return nil, 0, fmt.Errorf("couldn't find kfunc '%s' in modules: %v: %w", kfunc.Name, err, ErrNotSupported)
		}

		if err := btf.CheckTypeCompatibility(kfunc.Type, fn.Type); err != nil {
			return nil, 0, err
		}

		return module, id, nil
	}
	if err != nil {
		return nil, 0, fmt.Errorf("couldn't find kfunc '%s' in kernel spec: %v: %w", kfunc.Name, err, ErrNotSupported)
	}

	// we found the kfunc in vmlinux
	if err := btf.CheckTypeCompatibility(kfunc.Type, fn.Type); err != nil {
		return nil, 0, err
	}

	id, err := kernelSpec.TypeID(fn)
	if err != nil {
		return nil, 0, err
	}
	return nil, id, nil
}
```

  And then call this function within `fixupKfuncs()`. Currently `*btf.Handle.Spec()` makes this approach very slow but @lmb should fix that soonish.
* I've made a comment regarding camel-case variable naming, please make sure you name all your variables according to golang naming conventions.
* How we deal with the FdArray is probably something for round 2.

**Comment by shun159** [comment]
Thank you for your review. I have made some fixes as follows:
- Revert elf_reader (I havec confirmed that the Spec() is improved, pretty good)
- Fix `fixupKfunc` to return a list of FD
- Fix `testdata/kfunc.c` to add a kernel module ksym
- Add `LoadModuleHandle` that returns a list of module `Handle`

**Comment by shun159** [comment]
I have made some tentative changes. Specifically, I added a new field to the `Program` struct called `modBtfHandles`. This field contains a list of `*btf.Handle` and is utilized to close file descriptors when `(*Program).Close()` is called.

File: elf_reader.go [hunk 2]
```diff
@@ -22,17 +24,33 @@ const kconfigMap = ".kconfig"
 
 type kfuncMeta struct{}
 
+type kfuncDesc struct {
+	ksym_name   string
```
[the remark this answers is not in the document]
Comment (shun159) [reply elf_reader.go]: Thanks. I removed codes from `elf_reader`, and I stick to the naming convention.

File: linker.go [hunk 3]
```diff
@@ -228,27 +222,13 @@ fixups:
 		}
 
 		// check meta, if no meta return err
-		kfm, _ := ins.Metadata.Get(kfuncMeta{}).(*btf.Func)
+		kfm, _ := ins.Metadata.Get(kfuncMeta{}).(*kfuncDesc)
 		if kfm == nil {
 			return fmt.Errorf("kfunc call has no kfuncMeta")
 		}
 
-		var fn *btf.Func
```
[the remark this answers is not in the document]
Comment (shun159) [reply linker.go:238]: :white_check_mark:

File: testdata/kmod_kfunc.c [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#include "common.h"
+
+char __license[] __section("license") = "Dual MIT/GPL";
+
+extern void bpf_testmod_test_mod_kfunc(int) __ksym;
+extern void bpf_kfunc_call_test_mem_len_pass1(void *mem, int len) __ksym;
+
+__section("tc") int call_kmod_kfunc() {
+    bpf_testmod_test_mod_kfunc(0);
```
[the remark this answers is not in the document]
Comment (shun159) [reply testdata/kmod_kfunc.c]: :white_check_mark:

File: prog.go [hunk 5]
```diff
@@ -252,6 +256,11 @@ func newProgramWithOptions(spec *ProgramSpec, opts ProgramOptions) (*Program, er
 		attr.LineInfoRecSize = btf.LineInfoSize
 		attr.LineInfoCnt = uint32(len(lib)) / btf.LineInfoSize
 		attr.LineInfo = sys.NewSlicePointer(lib)
+
+		if spec.BtfHandles != nil {
```
[the remark this answers is not in the document]
Comment (shun159) [reply prog.go]: :white_check_mark:

File: btf/btf.go [hunk 6]
```diff
@@ -306,6 +306,32 @@ var kernelBTF struct {
 	fallback bool
 }
 
+type ModuleSpec struct {
```
[the remark this answers is not in the document]
Comment (shun159) [reply btf/btf.go]: :white_check_mark:

File: linker.go (hunk 1 above)
Comment (shun159) [reply linker.go:215]: :white_check_mark:

File: linker.go [hunk 7]
```diff
@@ -190,17 +190,20 @@ func fixupAndValidate(insns asm.Instructions) error {
...
 }
 
 // fixupKfuncs loops over all instructions in search for kfunc calls.
 // If at least one is found, the current kernels BTF is loaded to set Instruction.Constant
 // to the running kernels btf id of the btf.Func in the instructions Metadata for all kfunc call instructions.
-func fixupKfuncs(insns asm.Instructions) error {
+func fixupKfuncs(insns asm.Instructions) ([]*btf.Handle, error) {
+	fdArray := []*btf.Handle{}
+
+	modHandles, err := btf.LoadModuleHandle()
```
Comment (shun159) [inline linker.go:202]: Let me refactor this function.

**Review by rgo3 — CHANGES_REQUESTED** [review]
Nice Progress!
As promised for round two, this (or at least in similar fashion depending on a review by @lmb) is how I would deal with the fdArray:

```go
// handles stores handle objects to avoid gc cleanup
type handles []*btf.Handle

func (hs handles) add(h *btf.Handle) (int, error) {
	if h == nil {
		return 0, nil
	}

	if len(hs) == math.MaxInt16 {
		return 0, fmt.Errorf("can't add more than %d module FDs to fdArray", math.MaxInt16)
	}

	hs = append(hs, h)

	// return length of slice so that indexes start at 1
	return len(hs), nil
}

func (hs handles) fdArray() []int32 {
	// first element of fda is reserved as no module can be indexed with 0
	fda := []int32{0}
	for _, h := range hs {
		fda = append(fda, int32(h.FD()))
	}

	return fda
}

func (hs handles) close() {
	for _, h := range hs {
		h.Close()
	}
}
```

With this you can have `func fixupKfuncs(...) (handles, error)` and it lets you keep references to all the necessary module handles so that they don't get cleaned up by the GC, it provides a routine to call `defer hs.close()` once `hs, err := fixupKfuncs(...)` returns successfully, and keeps all the logic to calculate the correct index in the fdArray in one place as well without having to change too many other things.
I guess a downside of this approach is that multiple FDs pointing to the same module could be added to the fdArray, something you are trying to avoid with the usage of a map, but I think optimizing this is probably a good follow-up task.

Drive-by observation, we should add something like the following to TestKfunc:
```go
haveTestmod := false
if !testutils.IsKernelLessThan(t, "5.11") {
	// See https://github.com/torvalds/linux/commit/290248a5b7d829871b3ea3c62578613a580a1744
	testmod, err := btf.FindHandle(func(info *btf.HandleInfo) bool {
		return info.IsModule() && info.Name == "bpf_testmod"
	})
	if err != nil && !errors.Is(err, btf.ErrNotFound) {
		t.Fatal(err)
	}
	haveTestmod = testmod != nil
	testmod.Close()
}
if !haveTestmod {
	t.Skip("bpf_testmod not loaded"
})
```
Otherwise this test will fail if people run the tests locally.

File: linker.go (hunk 7 above)
Comment (shun159) [reply linker.go]: Refactored

File: btf/btf.go [hunk 8]
```diff
@@ -893,3 +893,25 @@ func probeBTF(typ Type) error {
 
 	return err
 }
+
+// LoadModuleHandle returns a list of module BTF handle.
```
[the remark this answers is not in the document]
Comment (shun159) [reply btf/btf.go]: I agree. So, I removed the function.

**Comment by rgo3** [comment]
I think I'd just put it in the file where `fixupKfuncs` and `findKfuncInKernel` for now, so linker.go?

**Comment by shun159** [comment]
I made a small change to avoid adding redundant FD (6c0d10b)
If this change is not reasonable way, I can revert the commit

File: linker.go [hunk 9]
```diff
@@ -190,17 +242,15 @@ func fixupAndValidate(insns asm.Instructions) error {
...
-
 	return nil
 }
 
 // fixupKfuncs loops over all instructions in search for kfunc calls.
 // If at least one is found, the current kernels BTF is loaded to set Instruction.Constant
 // to the running kernels btf id of the btf.Func in the instructions Metadata for all kfunc call instructions.
-func fixupKfuncs(insns asm.Instructions) error {
+func fixupKfuncs(insns asm.Instructions) (*handlesMap, error) {
+	fdArray := newHandlesMap()
```
Comment (rgo3) [inline linker.go:252]: I'd move the declaration of `fdArray` to before the for loop once we actually find kfuncs. In error cases you can return `nil` for `handles`.

**Review by rgo3 — CHANGES_REQUESTED** [review]
Apologies for the delay @shun159, this fell of my plate last week...

I've made a couple more comments, but this is getting in good shape.
CI is still unhappy with the elfs of kfunc.c, I think you need to recompile them with `make container-all`. Feel free to squash the commits into the first one, since that commit is signed this will also satisfy the DCO check.

Once CI is happy and the leftover feedback is addressed I am happy to remove the draft status from the PR and we can get a review from a maintainer as well.

File: elf_reader_test.go [hunk 10]
```diff
@@ -646,6 +646,21 @@ func TestKfunc(t *testing.T) {
 			t.Fatalf("Expected kfunc to return value 1, got %d", ret)
 		}
 
+		haveTestmod := false
```
[the remark this answers is not in the document]
Comment (shun159) [reply elf_reader_test.go]: :ballot_box_with_check:

File: linker.go [hunk 11]
```diff
@@ -5,12 +5,64 @@ import (
...
 	"io"
+	"math"
 
 	"github.com/cilium/ebpf/asm"
 	"github.com/cilium/ebpf/btf"
 	"github.com/cilium/ebpf/internal"
 )
 
+// handles stores handle objects to avoid gc cleanup
+type handlesMap struct {
```
[the remark this answers is not in the document]
Comment (shun159) [reply linker.go]: :ballot_box_with_check:

File: prog.go [hunk 12]
```diff
@@ -164,6 +165,7 @@ type Program struct {
 	name       string
 	pinnedPath string
 	typ        ProgramType
+	handles    *handlesMap
```
[the remark this answers is not in the document]
Comment (shun159) [reply prog.go]: :ballot_box_with_check:

File: linker.go (hunk 9 above)
Comment (shun159) [reply linker.go]: :ballot_box_with_check:

File: linker.go [hunk 13]
```diff
@@ -232,32 +282,31 @@ fixups:
...
-			return fmt.Errorf("couldn't resolve %s in kernel spec: %v: %w", kfm.Name, err, ErrNotSupported)
+			return fdArray, fmt.Errorf("kfunc call has no kfuncMeta")
 		}
 
-		if err := btf.CheckTypeCompatibility(kfm.Type, fn.Type); err != nil {
-			return err
-		}
+		idx := 0
+		kfuncHandle, id, err := findKfuncInKernel(kernelSpec, kfm)
+		if kfuncHandle != nil {
```
[the remark this answers is not in the document]
Comment (shun159) [reply linker.go]: :ballot_box_with_check:

File: linker.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (shun159) [reply linker.go]: >Can multiple modules have the same name while not including the same kfuncs? If that's a possibility the current approach would introduce a bug where we find a kfunc in a module but then don't use the actual handle and just point to a previously found one with the same name. 

I agree, that's possible. I reverted the change.

File: elf_reader_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (shun159) [reply elf_reader_test.go]: I'm sorry. I didn't understand your feedback. I have fixed.

**Review by rgo3 — COMMENTED** [review]
Nice work, just a couple more minor comments. I'll remove the draft status from the PR and will add @lmb as a second reviewer.

File: linker.go [hunk 14]
```diff
@@ -190,17 +225,14 @@ func fixupAndValidate(insns asm.Instructions) error {
...
-	}
-
 	return nil
 }
 
 // fixupKfuncs loops over all instructions in search for kfunc calls.
 // If at least one is found, the current kernels BTF is loaded to set Instruction.Constant
 // to the running kernels btf id of the btf.Func in the instructions Metadata for all kfunc call instructions.
-func fixupKfuncs(insns asm.Instructions) error {
+func fixupKfuncs(insns asm.Instructions) (*handles, error) {
```
[the remark this answers is not in the document]
Comment (shun159) [reply linker.go]: :ballot_box_with_check:

File: linker.go [hunk 15]
```diff
@@ -209,15 +241,16 @@ func fixupKfuncs(insns asm.Instructions) error {
...
 fixups:
 	// only load the kernel spec if we found at least one kfunc call
-	s, err := btf.LoadKernelSpec()
+	kernelSpec, err := btf.LoadKernelSpec()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	fdArray := &handles{}
```
[the remark this answers is not in the document]
Comment (shun159) [reply linker.go]: :ballot_box_with_check:

File: prog.go [hunk 16]
```diff
@@ -268,6 +269,18 @@ func newProgramWithOptions(spec *ProgramSpec, opts ProgramOptions) (*Program, er
 		return nil, err
 	}
 
+	handles, err := fixupKfuncs(insns)
+	if err != nil {
+		return nil, fmt.Errorf("fixing up kfuncs: %w", err)
+	}
+
+	if handles != nil {
```
[the remark this answers is not in the document]
Comment (shun159) [reply prog.go]: :ballot_box_with_check:

File: prog.go [hunk 17]
```diff
@@ -1000,3 +1013,38 @@ func findTargetInProgram(prog *Program, name string, progType ProgramType, attac
...
+// Returns btf.ErrNotFound if the target can't be found in vmlinux and any kernel module.
+func findKfuncInKernel(kernelSpec *btf.Spec, kfunc *btf.Func) (*btf.Handle, btf.TypeID, error) {
+	var fn *btf.Func
+	err := kernelSpec.TypeByName(kfunc.Name, &fn)
+	if errors.Is(err, btf.ErrNotFound) {
+		// couldn't find kfunc in vmlinux, searching modules.
+		module, id, err := findTargetInModule(kernelSpec, kfunc.Name, kfunc)
+		if err != nil {
+			return nil, 0, fmt.Errorf("couldn't find kfunc '%s' in modules: %v: %w", kfunc.Name, err, ErrNotSupported)
+		}
```
[the remark this answers is not in the document]
Comment (shun159) [reply prog.go:1030]: :ballot_box_with_check:

File: prog.go [hunk 18]
```diff
@@ -1000,3 +1013,38 @@ func findTargetInProgram(prog *Program, name string, progType ProgramType, attac
 
 	return spec.TypeID(targetFunc)
 }
+
+// find an attach target type in a kernel module and the vmlinux
```
[the remark this answers is not in the document]
Comment (shun159) [reply prog.go]: :ballot_box_with_check:

File: prog.go [hunk 19]
```diff
@@ -1000,3 +1013,38 @@ func findTargetInProgram(prog *Program, name string, progType ProgramType, attac
...
+// Returns btf.ErrNotFound if the target can't be found in vmlinux and any kernel module.
```
[the remark this answers is not in the document]
Comment (shun159) [reply prog.go]: :ballot_box_with_check:

**Comment by shun159** [comment]
> Nice work, just a couple more minor comments. I'll remove the draft status from the PR and will add @lmb as a second reviewer.

Thank you for your feedback! I fixed.

File: linker.go [hunk 20]
```diff
@@ -232,32 +265,29 @@ fixups:
...
-
-		var fn *btf.Func
-		if err := s.TypeByName(kfm.Name, &fn); err != nil {
-			return fmt.Errorf("couldn't resolve %s in kernel spec: %v: %w", kfm.Name, err, ErrNotSupported)
+			return nil, fmt.Errorf("kfunc call has no kfuncMeta")
 		}
 
-		if err := btf.CheckTypeCompatibility(kfm.Type, fn.Type); err != nil {
-			return err
+		idx := 0
```
Comment (lmb) [inline linker.go:271]: Nit: is it necessary to declare `idx` here?

File: prog.go [hunk 21]
```diff
@@ -1000,3 +1013,41 @@ func findTargetInProgram(prog *Program, name string, progType ProgramType, attac
...
+		module, id, err := findTargetInModule(kernelSpec, kfunc.Name, kfunc)
+		if errors.Is(err, btf.ErrNotFound) {
+			return nil, 0, fmt.Errorf("kfunc '%s' doesn't exist in vmlinux or modules: %v: %w", kfunc.Name, err, btf.ErrNotFound)
+		}
+
+		if err != nil {
+			return nil, 0, fmt.Errorf("couldn't find kfunc '%s' in modules: %v: %w", kfunc.Name, err, ErrNotSupported)
+		}
+
+		if err := btf.CheckTypeCompatibility(kfunc.Type, kfunc.Type); err != nil {
```
Comment (lmb) [inline prog.go:1032]: Would it be possible to invoke this in `fixupKfuncs` instead? That would reduce some duplication.

File: testdata/kfunc.c [hunk 22]
```diff
@@ -2,9 +2,11 @@
 
 char __license[] __section("license") = "Dual MIT/GPL";
 
+extern void bpf_testmod_test_mod_kfunc(int) __ksym;
```
Comment (lmb) [inline testdata/kfunc.c:5]: Could you split this into a separate file `kfunc-kmod.c` plus a separate test `TestKfuncKmod` or similar? Then we still test regular kfuncs on systems that don't have bpf_testmod.

**Review by lmb — CHANGES_REQUESTED** [review]
Looks great, thanks @shun159 and @rgo3! Some nits, but this is pretty much good to go.

File: testdata/kfunc.c (hunk 22 above)
Comment (shun159) [reply testdata/kfunc.c]: Fixed :ballot_box_with_check:

File: linker.go (hunk 20 above)
Comment (shun159) [reply linker.go]: Fixed :ballot_box_with_check:

File: prog.go (hunk 21 above)
Comment (shun159) [reply prog.go]: Fixed :ballot_box_with_check: 
how about this?

File: elf_reader_test.go [hunk 23]
```diff
@@ -615,6 +615,23 @@ func TestKconfig(t *testing.T) {
 
 func TestKfunc(t *testing.T) {
 	testutils.SkipOnOldKernel(t, "5.18", "bpf_kfunc_call_test_mem_len_pass1")
+
+	haveTestmod := false
+	if !testutils.IsKernelLessThan(t, "5.11") {
```
[the remark this answers is not in the document]
Comment (shun159) [reply elf_reader_test.go]: `import cycle not allowed` warning appeared when this put into testutils. I'm wondering how avoid circular dependency.

[the remark this answers is not in the document]
Comment (lmb) [reply elf_reader_test.go]: Just keep it in `helpers_test.go` in the ebpf package.

[the remark this answers is not in the document]
Comment (shun159) [reply elf_reader_test.go]: I added `helpers_test.go` in the ebpf package.

**Review by lmb — APPROVED** [review]
Thanks for the hard work @shun159 and @rgo3!

File: prog.go (hunk 21 above)
Comment (rgo3) [reply prog.go]: Now the `btf.CheckTypeCompatibility` checks are wrong IMO. Now we call it in `findKfuncInKernel` and `fixupKfuncs`, and in `fixupKfuncs` we check `btf.CheckTypeCompatibility(kfm.Type, kfm.Type)` which will always pass.

Comment (shun159) [reply prog.go]: @rgo3, 
I'm sorry for the late reply, and for made a degrade. 
I have read your changes, thank you.

Comment (rgo3) [reply prog.go]: I'm still not super happy with what I added, but that way we only do the type compatibility check once. Let's discuss next week with Lorenz, we should be close to the finish line :muscle:

Comment (shun159) [reply prog.go]: 🙇

Comment (lmb) [reply prog.go]: I had to do some additional refactoring, but I think it ended up fine / not much more complex. PTAL @rgo3

**Review by lmb — APPROVED** [review]

**Review by rgo3 — APPROVED** [review]
:partying_face:

**Comment by shun159** [comment]
Thanks for your kindly support, @rgo3 and @lmb! :partying_face:

**Pull request merged**